### PR TITLE
refactor: remove optional comparison override

### PR DESCRIPTION
**Pull request by markusfassbender**
### Issue

Removes override of comparison operators `<` and `>` for optional values. How to handle optionality should be up to the specific use case. With this change the compiler works again as expected.

To give some convenience like `array.sort(by:)` there is a new generic func `OptionalComparison.prependingNilAscending(lhs:rhs:). One needs to use it explicitly instead of a hidden operator override.

Review `OptionalComparisonTests.swift` first for exampels of old and new behaviour.

### Testing

- Run unit tests suite, passed all suites [`Test Development`](https://github.com/wireapp/wire-ios/actions/runs/9002732692) ✅
- I think it's not possible to test all cases in runtime explicitly, but we have regression tests.

---

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [ ] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

### CODE REVIEW

File: wire-ios/Wire-iOS/Sources/UserInterface/Calling/Helper/Array+CallParticipant.swift [hunk 1]
```diff
@@ -22,13 +22,19 @@ import WireSyncEngine
...
     var hasMoreThanTwoConnectedParticipants: Bool {
-        return filter(\.state.isConnected).count > 2
+        filter(\.state.isConnected).count > 2
     }
 
     mutating func sortByName(selfStreamId: AVSClient?) {
-        self = self.sorted {
-            $0.streamId == selfStreamId ||
-                $0.user.name?.lowercased() < $1.user.name?.lowercased()
+        sort {
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Calling/Helper/Array+CallParticipant.swift:29]: I think `sort` us equal in the result and using `self = self.sorted` just additionally copies arrays around.

File: wire-ios/Wire-iOS/Sources/UserInterface/SelfProfile/AccountSelectorController.swift [hunk 2]
```diff
@@ -46,8 +53,14 @@ final class AccountSelectorController: UIViewController {
...
-        let showAccounts = SessionManager.shared?.accountManager.accounts.count > 1
-        guard showAccounts != showAccounts else { return }
+        let showAccounts: Bool
+        if let count = SessionManager.shared?.accountManager.accounts.count, count > 1 {
+            showAccounts = true
+        } else {
+            showAccounts = false
+        }
+
+        guard showAccounts != self.showAccounts else { return }
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/SelfProfile/AccountSelectorController.swift]: This `self` is a bugfix as discussed in the `iOS developers` chat in Wire.

**Comment by markusfassbender** [comment]
> ### Issue
> Removes override of comparison operators `<` and `>` for optional values. How to handle optionality should be up to the specific use case.
> 
> To give some convenience like `array.sort(by:)` there is a new generic func `OptionalComparison.prependingNilAscending(lhs:rhs:). One needs to use it explicitly instead of a hidden operator override.
> 
> ### Testing
> * Build
> * Run unit tests suite
> * I think it's not possible to test all cases in runtime explicitly
> 
> ### Checklist
> * [ ]  Title contains a reference JIRA issue number like `[WPB-XXX]`.
> * [ ]  Description is filled and free of optional paragraphs.
> * [ ]  Adds/updates automated tests.

File: wire-ios-utilities/Source/Public/OptionalComparison.swift [hunk 3]
```diff
@@ -18,17 +18,31 @@
...
+    ///
+    /// - Parameters:
+    ///   - lhs: An optional comparable value on the left hand side.
+    ///   - rhs: An optional comparable value on  the right hand side.
+    ///
+    /// - Returns:
+    ///  `true` if left hand side is smaller than right hand side.
+    /// `nil` is  considered as smaller than any value.
+    /// If both paratemeters are `nil`, the method returns `false`.
+    public static func prependingNilAscending<T: Comparable>(lhs: T?, rhs: T?) -> Bool {
```
Comment (markusfassbender) [inline wire-ios-utilities/Source/Public/OptionalComparison.swift:32]: The naming can be discussed.. it was not easy for me to express what it does in a comprehensive way. Thanks also to @caldrian for the proposal.

File: wire-ios-utilities/Source/Public/OptionalComparison.swift [hunk 4]
```diff
@@ -18,17 +18,31 @@
...
-    case (nil, _?): return true
-    default: return false
-    }
-}
+public enum OptionalComparison {
+    /// Compares two optional values in ascending order.
+    ///
+    /// - Parameters:
+    ///   - lhs: An optional comparable value on the left hand side.
+    ///   - rhs: An optional comparable value on  the right hand side.
```
Comment (caldrian) [inline wire-ios-utilities/Source/Public/OptionalComparison.swift]:
```suggestion
    ///   - rhs: An optional comparable value on the right hand side.
```

File: wire-ios-utilities/Source/Public/OptionalComparison.swift (hunk 3 above)
Comment (caldrian) [reply wire-ios-utilities/Source/Public/OptionalComparison.swift:32]: I actually thought about an extension to `Collection`, something like this:

```
extension Collection {

    func sortedAscendingPrependingNil<Wrapped>() -> [Element] where Element == Optional<Wrapped>, Wrapped: Comparable {
        sorted { lhs, rhs in

            if lhs == nil, rhs == nil {
                return false
            }

            guard let lhs else {
                return true
            }

            guard let rhs else {
                return false
            }

            return lhs < rhs
        }
    }
}
```

Comment (caldrian) [reply wire-ios-utilities/Source/Public/OptionalComparison.swift:32]: Alternatively, to keep the syntax similar and let the user provide a closure:
```
extension Collection {

    func sortedPrependingNil<Wrapped>(
        by areInIncreasingOrder: (Wrapped, Wrapped) throws -> Bool
    ) rethrows -> [Element]
    where Element == Wrapped?, Wrapped: Comparable {
        try sorted { lhs, rhs in

            if lhs == nil, rhs == nil {
                return false
            }

            guard let lhs else {
                return true
            }

            guard let rhs else {
                return false
            }

            return try areInIncreasingOrder(lhs, rhs)
        }
    }
}
```

**Review by typfel — APPROVED** [review]

File: wire-ios-utilities/Source/Public/OptionalComparison.swift [hunk 5]
```diff
@@ -18,17 +18,31 @@
...
+    ///
+    /// - Returns:
+    ///  `true` if left hand side is smaller than right hand side.
+    /// `nil` is  considered as smaller than any value.
+    /// If both paratemeters are `nil`, the method returns `false`.
```
Comment (netbe) [inline wire-ios-utilities/Source/Public/OptionalComparison.swift]:
```suggestion
    /// If both parameters are `nil`, the method returns `false`.
```

File: wire-ios-utilities/Source/Public/OptionalComparison.swift (hunk 3 above)
Comment (netbe) [reply wire-ios-utilities/Source/Public/OptionalComparison.swift:32]: I am not sure to grasp what we gain compared to the custom operator? clarity?

Comment (markusfassbender) [reply wire-ios-utilities/Source/Public/OptionalComparison.swift:32]: @netbe the reason why this change is important is, that one can not find a general solution to the question "how to compare nil with a value".

But with the custom operator it implicitly "works" without any compiler warning, which can lead to interesting cases like here:

1. Does `epoch=nil` enter the if with `true` or `false`

Screenshot 2024-05-14 at 16 41 29

2. What if on of the dates is nil? what if both values are `nil`?
Screenshot 2024-05-14 at 16 46 39

Comment (caldrian) [reply wire-ios-utilities/Source/Public/OptionalComparison.swift:32]: Nice, `sortedAscendingPrependingNil` is very clear about the handling of `nil`.

**Review by caldrian — APPROVED** [review]

**Pull request merged**